Update Migrate FAQ based on user issue discussion (#2377)
prisma/prisma-engines: migration-engine/ARCHITECTURE.md

  by the bad deployment.
- The down migration actually works. Are your down migrations tested?

> In short, down migrations give you a sense of security, but it is often a false

Reviewer: Oh so opinionated 😆

Author: Migrate itself is a very opinionated tool — it's also incomplete, but hiding the opinionation makes it seem even more incomplete imo.

  looking at the `_prisma_migrations` table in the target database.
- Run the migrations that have not been applied yet, in chronological order.

`prisma migrate deploy` is the command meant to be used for _unattended_
migrations (as part of CI). As such, it should be as reliable, predictable and
deterministic as possible.

### Why does `prisma migrate deploy` not warn when a migration is in the migrations table but not in the migrations folder?

We don't want to warn if an already applied migration is missing from the
migration directory because it would prevent squashing migrations and

> deployment from not-properly-rebased branches.

Reviewer: OT: Is squashing migrationssomething we support with a workflow or a manual thing?

Author: It's something we don't communicate in the docs, but there is no reason we shouldn't, it works and we have tests in the migration engine checking that it does.

Author: Product never picked it up as a required feature, but the general idea was to not communicate about it until it's gotten product design effort put into it.

> In general, by design, `deploy` errs on the side of not standing in the way of

Reviewer 2: ```suggestion
In general, by design, `deploy` errors on the side of not standing in the way of
```

> deploying your migrations.

Reviewer 2: Can we use another word for `errs` I first thought it was a typo for errors 😓

Author: It's a rather common expression. I can't easily find synonyms (https://www.merriam-webster.com/dictionary/err%20on%20the%20side%20of doesn't list any), but I'm happy to oblige if there's one.

Reviewer 2: ```suggestion
In general, by design, `deploy` takes the safe approach of not standing in the way of
deploying your migrations.
```
maybe works too

Author: I'm trying hard to come up with a different phrasing for this. The best I have is "deploy is designed to not stand in the way of deploying your migration", but it's not exactly what I mean. There's a tradeoff and there is no right choice and no safe approach, we have to balance our design, and the balance tends to go a bit more on the "deploy anything" than the "all checks enabled in strictest mode" direction.

Reviewer 2: The currently merged doc with `errs` doesn't help to understand this though

### Why does `prisma migrate deploy` not detect drift?

We don't detect drift because we want to keep the deployment path light and

> simple, and because we would need a shadow database for that. Many people would

Reviewer: There would also be other ways to detect a mismatch of production and the "before" state of the database the migration was created on if Migrate worked differently and had other information. This statement sounds a bit like a shadow database would be the only way, but comparing an introspected schema of the production database with a "before" schema (if that was available) would for example also be a way.

Maybe qualify this sentence better to show it means "the way Prisma Migrate currently works".

Author: >  comparing an introspected schema of the production database with a "before" schema (if that was available

The only way we have to obtain that "before" schema is exactly the shadow database. I don't see a reasonable alternative with the current architecture. I can qualify.

Reviewer: Oh yeah, I know that how Migrate works this is not possible any other way. But that is just because how Prisma Migrate works, not universal truth 👍

Author: Totally!

not be comfortable with creating/using temporary databases being on the
deployment path. And as stated in the previous question, in general, by design,

> `deploy` errs on the side of not standing in the way of deploying your

Reviewer 2: ```suggestion
`deploy` errors on the side of not standing in the way of deploying your
```

> `deploy` errs on the side of not standing in the way of deploying your

Reviewer 2: ```suggestion
`deploy` takes the safe approach of not standing in the way of deploying your
```

migrations.

### What happens when a migration fails in `prisma migrate deploy`?

> If a migration fails during deployment, you will see the error. Then `prisma

Reviewer: Semi OT: Can we create an example, and possibly documentation, for both the error case of a migratoin failing in `deploy` and then how to handle the cleanup?

Author: We have public docs with an example, I think. I'll have to look for the docs page.

Author: https://www.prisma.io/docs/guides/database/production-troubleshooting

Reviewer: (That links def should have "migrate" or "migrations" in the URL)

migrate status` and `prisma migrate deploy` commands (in subsequent runs) will
show you the failed state with the error message when you run them again.

> It's then your responsibility to fix what failed, and marked the migration as

Reviewer: scary 😨

Author: Yes, that's why I've been mentioning every occasion I get (it's good to put it in this FAQ to emphasize, it's really something I repeat all the time on slack). It's the biggest UX gap we have.

> It's then your responsibility to fix what failed, and marked the migration as

Reviewer: And fix:
```suggestion
It's then your responsibility to fix what failed, and mark the migration as
```

Author: :+1:

applied or rolled back with `prisma migrate resolve`, so you can deploy
migrations again.

Prisma Migrate does not offer much help at that last stage, but this is

> something we are working on defining and prioritizing.

Reviewer: Issue link?

Author: We had issues that were closed. I'll open a new placeholder one.

Author: Added a link to the new issue and pushed.
